evidence/reactor.go

```diff
@@ -21,233 +46,346 @@ const (
+// envelopes on each. In addition, it also listens for peer updates and handles
+// messages on that p2p channel accordingly. The caller must be sure to execute
+// OnStop to ensure the outbound p2p Channels are closed. No error is returned.
+func (r *Reactor) OnStart() error {
+	go r.processEvidenceCh()
+	go r.processPeerUpdates()
+
+	return nil
 }
 
-// AddPeer implements Reactor.
-func (evR *Reactor) AddPeer(peer p2p.Peer) {
-	go evR.broadcastEvidenceRoutine(peer)
+// OnStop stops the reactor by signaling to all spawned goroutines to exit and
+// blocking until they all exit.
+func (r *Reactor) OnStop() {
+	// Close closeCh to signal to all spawned goroutines to gracefully exit. All
+	// p2p Channels should execute Close().
+	close(r.closeCh)
+
+	// Wait for all spawned peer evidence broadcasting goroutines to gracefully
+	// exit.
+	r.peerWG.Wait()
+
+	// Wait for all p2p Channels to be closed before returning. This ensures we
+	// can easily reason about synchronization of all p2p Channels and ensure no
+	// panics will occur.
+	<-r.evidenceCh.Done()
+	<-r.peerUpdates.Done()
 }
 
-// Receive implements Reactor.
-// It adds any received evidence to the evpool.
-// XXX: do not call any methods that can block or incur heavy processing.
-// https://github.com/tendermint/tendermint/issues/2888
-func (evR *Reactor) Receive(chID byte, src p2p.Peer, msgBytes []byte) {
-	evis, err := decodeMsg(msgBytes)
-	if err != nil {
-		evR.Logger.Error("Error decoding message", "src", src, "chId", chID, "err", err)
-		evR.Switch.StopPeerForError(src, err)
-		return
+// handleEvidenceMessage handles enevelopes sent from peers on the EvidenceChannel.
+// It returns an error only if the Envelope.Message is unknown for this channel
+// or if the given evidence is invalid. This should never be called outside of
+// handleMessage.
+func (r *Reactor) handleEvidenceMessage(envelope p2p.Envelope) error {
+	switch msg := envelope.Message.(type) {
+	case *tmproto.EvidenceList:
+		r.Logger.Debug("received evidence list", "num_evidence", len(msg.Evidence), "peer", envelope.From.String())
+
+		// TODO: Refactor the Evidence type to not contain a list since we only ever
+		// send and receive one piece of evidence at a time. Or potentially consider
+		// batching evidence.
+		for i := 0; i < len(msg.Evidence); i++ {
+			ev, err := types.EvidenceFromProto(&msg.Evidence[i])
+			if err != nil {
+				r.Logger.Error("failed to convert evidence", "peer", envelope.From.String(), "err", err)
+				continue
+			}
+
+			if err := r.evpool.AddEvidence(ev); err != nil {
+				r.Logger.Error("failed to add evidence", "peer", envelope.From.String(), "err", err)
+
+				// If we're given invalid evidence by the peer, notify the router that
+				// we should remove this peer by returning an error.
+				if _, ok := err.(*types.ErrInvalidEvidence); ok {
+					return err
+				}
+			}
+		}
+
+	default:
+		r.Logger.Error("received unknown message", "msg", msg, "peer", envelope.From.String())
```

Comment:
This also gets logged in `processEvidenceChannel`, and should perhaps not even be logged there but rather in the shim?

Reply:
There is no `processEvidenceChannel`. This is the only log that logs an unknown message.

Reply:
Meant `processEvidenceCh`. We log this message `received unknown message` here, and also return an error `received unknown message` on the line below. That returned error is caught and logged here:

https://github.com/tendermint/tendermint/blob/d791336c0331e4477839f019d12a70d556b886f9/evidence%2Freactor.go#L190-L191

Reply:
Ahh yes, I've tweaked the logging a bit.